add code generator scripts and firewall CRD

sugangli opened the pull request:
Copy types.go from the gke-node-firewall/pkg/apis/gcpfirewall, and generate the files locally.

sugangli commented:
/ok-to-test

MrHohn commented:
/ok-to-test

cici37 commented:
cloud-provider-gcp-e2e-full is currently blocked by https://github.com/kubernetes/cloud-provider-gcp/pull/273

cici37 commented:
#273 is merged. Please fetch the latest changes which will unblock e2e failures. Thanks

sugangli commented:
/retest

cici37 commented on go.mod:
```diff
@@ -4,37 +4,25 @@ go 1.16
 
```
Those deletions in go.mod could be separate from the current pr and should not fail gcp-e2e-all. :)

freehan reviewed (commented):
can you split into multiple commits?

one for manual changes and one for auto code gen.

sugangli commented:
/retest

freehan commented on pkg/apis/gcpfirewall/v1alpha1/types.go:
```diff
@@ -0,0 +1,181 @@
+	// Name of the firewall rule that's configured in GCP.
+	// For classic GCP firewalls, this field needs to be unique. For FW 3.0,
+	// this field will be in the rule description.
+	Name string `json:"name"`
```
I am wondering if you could just reuse the name of the GCPFirewall object as that one is unique within the API server. Why do you need another name in the CR?

sugangli replied:
This CR will have multiple instances in the API server. Just for ingress controller, there will be two: one for L7 firewall, another one for L4 firewall. GCP firewall controller manages all the firewall CRs.

skmatti replied:
I agree with @freehan, we can remove this field now. We added this field as the previous version of the spec can define multiple GCE rules. 

@sugangli Can you clarify you comment?. Each GCE firewall that Ingress/ Service controllers wants to create is a separate GCPFirewall CR. So, there is a one-to-one map between GCPFirewall CR and the associated GCE firewall rule.

sugangli replied:
I thought it was the name of the GCPFirewall CR. In that case, we can remove it. How about the name in the GCPFirewallStatus?

skmatti replied:
I think we should keep the name in GCPFirewallStatus because GCPFirewall controller may choose take control of naming scheme in future

sugangli replied:
Updated.

freehan commented on pkg/apis/gcpfirewall/v1alpha1/types.go:
```diff
@@ -0,0 +1,181 @@
+	// +optional
+	// +kubebuilder:validation:Enum=ALLOW
+	// +kubebuilder:default=ALLOW
+	Action string `json:"action"`
```
do you want to include a few string constant to capture the valid input here?

skmatti replied:
Right now, ALLOW is only action possible. We might extend this in future to include DENY. This can be considered readonly for now as this field can be left empty.

sugangli replied:
Updated.

sugangli commented on hack/verify-codegen.sh:
```diff
@@ -0,0 +1,48 @@
+#!/bin/bash
```
SG. Will delete it.

skmatti commented on pkg/apis/gcpfirewall/v1alpha1/types.go:
```diff
@@ -0,0 +1,181 @@
+type Protocol string
+
+// CIDR defines a IP block.
+// TODO(b/186120065) Modify the validation to include IPv6 CIDRs with FW 3.0 support.
```
Remove internal bug references, replace this with your github id

sugangli replied:
Done.

freehan commented:
Can you reorganize the commits? Squash the ones that are addressing comments

sugangli commented:
> Can you reorganize the commits? Squash the ones that are addressing comments
They are in "address comments" (0d7fbbe) commit. The other commits are just adding the code-gen dependencies and updating the auto-generated yaml.

jiahuif commented:
Or you can ask tide to squash-merge this PR. The command is 
`/label tide/merge-method-squash`

sugangli commented:
> Or you can ask tide to squash-merge this PR, after approval of couse. The command is `/label tide/merge-method-squash`

Thx! Will try it next time!

freehan commented on pkg/apis/gcpfirewall/v1alpha1/types.go:
```diff
@@ -0,0 +1,186 @@
+
+	// Name of the GCP firewall rule in the GCE.
+	// +optional
+	Name string `json:"name"`
```
I am wondering if you should just include the selflink/URL of the firewall object directly. 

No need to have a specific name

freehan replied:
This would remove the need for specifying global vs. region as well.

skmatti replied:
Do we want to include project name/ number in the status ?. Also, the prefix googleapis.com keeps changing which might make the status uncertain. 

We have region vs global concept only for FW 3.0. With FW 3.0, firewall rules are not first class citizens in GCE anymore so they do not have self links. All GKE cluster rules will be a part of single firewall network policy resource.

A VPC firewall status will look like:
```
status:
  type: VPC
  name: gke-pfw-default-my-ingress
  conditions:
  - type: Enforced
     status: True
```

A global network firewall policy status will be:
```
status:
  type: GLOBAL
  priority: 10
  conditions:
  - type: Enforced
     status: True
```

A regional network firewall policy status will be:
```
status:
  type: REGIONAL
  priority: 10
  conditions:
  - type: Enforced
     status: True
```

skmatti replied:
If we are thinking about resource links, probably the following works:

```
status:
  type: VPCFirewall
  resource: global/firewalls/gke-pfw-default-my-ingress
  conditions:
  - type: Enforced
     status: True
```

For Global Network Firewall Policy:
```
status:
  type: GlobalNetworkFirewallPolicy
  resource: global/firewallPolicies/gke-pfw-default-my-ingress
  priority: 10
  conditions:
  - type: Enforced
     status: True
```

Does it look better than what is proposed right now?. I also wonder if we just want to keep the suffix part of resource link as this excludes GCP project info from the status. I feel including the full resource link might have some caveats (googleapis.com, compute alpha/beta/GA changes can modify the status)

freehan replied:
> With FW 3.0, firewall rules are not first class citizens in GCE anymore so they do not have self links. 

You mean FW 3.0 rule no longer has an object representation? I imagine all GCP api objects still has a resource URL right? 

> Also, the prefix googleapis.com keeps changing which might make the status uncertain.

This I agree. But we can counter that by having smarter util function for parsing.

freehan replied:
I am trying to point out that the Name and Global/Regional could not help pin point the actual GCP resource of FW rules.

skmatti replied:
I agree with confusion about VPC/Global/Regional. Maybe we can change the names to VPCFirewall/GlobalNetworkFirewallPolicy/RegionNetworkFirewallPolicy.

Also, we will include the resource suffixes (excluding compute and project part).

FW 3.0 has single object for all the firewall rules. Within the object, priority uniquely identifies the rule. So, there is no resource link for a firewall rule but there is one for the network firewall policy.

```
status:
  type: GlobalNetworkFirewallPolicy
  resource: global/firewallPolicies/gke-pfw-default-my-ingress
  priority: 10
  conditions:
  - type: Enforced
     status: True
```

Does this look like something you are recommending?. In the above status, resource is same for all firewall rules, the priority uniquely identifies the rule.

sugangli replied:
Updated according to our discussion today.

skmatti commented on pkg/apis/gcpfirewall/v1alpha1/types.go:
```diff
@@ -0,0 +1,182 @@
+}
+
+// GCPFirewallIngressPeer describes a peer to allow traffic from.
+type GCPFirewallIngressPeer struct {
```
```
// GCPFirewallIngressPeer describes a peer to allow traffic from. Items in the
// sub-fields are combined using a logical OR operation. If all the sub-fields
// are missing, then traffic is allowed from all sources.
type GCPFirewallIngressPeer struct {
    // IPBlocks specify the set of CIDRs that the rule applies to. If this field
    // is present and contains at least one item, this rule allows traffic only if
    // the traffic matches at least one item in the list.
    // Valid example list items are "192.168.1.1/24" or "2001:db9::/64".
    // +optional
    // +kubebuilder:validation:MinItems=1
    // +kubebuilder:validation:MaxItems=256
    IPBlocks []CIDR `json:"ipBlocks,omitempty"`
}
```

Adds more description and validation

sugangli replied:
done.

skmatti commented on pkg/apis/gcpfirewall/v1alpha1/types.go:
```diff
@@ -0,0 +1,187 @@
+
+const (
+	// FirewallRuleConditionReady indicates if the firewall rule is enforced.
+	FirewallRuleConditionReady FirewallRuleConditionType = "Ready"
```
FirewallRuleConditionEnforced FirewallRuleConditionType = "Enforced"

skmatti commented on pkg/apis/gcpfirewall/v1alpha1/types.go:
```diff
@@ -0,0 +1,187 @@
+	FirewallRuleReasonInvalid FirewallRuleConditionReason = "Invalid"
+
+	// FirewallRuleReasonGCPError is used if the sync fails due to a GCP error.
+	FirewallRuleReasonGCPError FirewallRuleConditionReason = "GCPError"
```
```
        // FirewallRuleReasonError is used if the sync fails due to an error.
	FirewallRuleReasonError FirewallRuleConditionReason = "Error"
```

sugangli replied:
Done.

freehan commented on pkg/apis/gcpfirewall/v1alpha1/types.go:
```diff
@@ -0,0 +1,190 @@
+	// +kubebuilder:default=ALLOW
+	Action Action `json:"action"`
+
+	// If set to true, the GCPFirewall is not synced by the controller.
```
nit: If the firewall rule is previously synced, it would be orphanded?

freehan replied:
from a consumer stand point. If I mark the CR as disabled, the Firewall controller would not act on the Firewall. Then it is effectively orphanded, right? Then whether the consumer (ingress-gce) decides to configure the firewall on its own and whether to GC the CR. It is orthogonal, no?

Also, in the case of FW 3.0. The FW rule will no longer implement this CR once marked disabled. In that case, it would not be orphanded, it would be removed. 

Am I understanding correctly?

freehan reviewed (commented):
LGTM overall
just one nit
